File: python/1d-dynamic-programming/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        maximum = 1
        minimum = 1

        answer = float('-inf')
        for i in range(len(nums)):
            if nums[i] != 0:
                maximum, minimum = max(maximum * nums[i], minimum * nums[i], nums[i]), min(maximum * nums[i],
                                                                                           minimum * nums[i], nums[i])
                answer = max(maximum, answer)
            else:
                answer = max(answer, 0)
                maximum = 1
                minimum = 1

        return answer

File: c++/1d-dynamic-programming/maximum-product-subarray.cc
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int answer = std::numeric_limits<int>::min();

        int maximum = 1;
        int minimum = 1;
        for (int num : nums) {
            if (num == 0) {
                maximum = 1;
                minimum = 1;
            } else {
                int old_maximum = maximum;
                maximum = std::max(maximum * num, minimum * num);
                minimum = std::min(minimum * num, old_maximum * num);

                answer = std::max(answer, maximum);
            }
            maximum = std::max(maximum, 1);
            answer = std::max(answer, num);
        }

        return answer;
    }
};
